usr/tgtd.c

```diff
@@ -519,6 +535,203 @@ static int parse_params(char *name, char *p)
 	return -1;
 }
 
+void *ha_heartbeat(void *arg)
+{
+	struct _ha_instance *hap = (struct _ha_instance *) arg;
+
+	while (1) {
+		pthread_setcancelstate(PTHREAD_CANCEL_DISABLE, 0);
+		ha_healthupdate(hap);
+		pthread_setcancelstate(PTHREAD_CANCEL_ENABLE, 0);
+		sleep(60);
+	}
+}
+
+enum tgt_svc_err {
+	TGT_ERR_INVALID_PARAM = 1,
+	TGT_ERR_TARGET_CREATE,
+	TGT_ERR_NO_DATA,
+	TGT_ERR_INVALID_JSON,
+	TGT_ERR_INVALID_TARGET_NAME,
+	TGT_ERR_INVALID_LUN_PATH,
+	TGT_ERR_INVALID_VMID,
+	TGT_ERR_INVALID_VMDKID,
+	TGT_ERR_LUN_CREATE,
+};
+
+static void set_err_msg(_ha_response *resp, enum tgt_svc_err err,
+	char *msg)
+{
+	char *err_msg = ha_get_error_message(ha, err, msg);
+	ha_set_response_body(resp, HTTP_STATUS_ERR, err_msg,
+		strlen(err_msg) + 1);
+	free(err_msg);
+}
+
+static int exec(char *cmd)
+{
+	FILE *filp = NULL;
+	int ret = 0;
+	int status = 0;
+
+	filp = popen(cmd, "r");
+	if (filp == NULL) {
+		return -1;
+	}
+
+	ret = pclose(filp);
+	status = WEXITSTATUS(ret);
+
+	return status;
+}
+
+static int target_create(const _ha_request *reqp,
+	_ha_response *resp, void *userp)
+{
+	char cmd[512];
+	const char *tid = ha_parameter_get(reqp, "tid");
+	int rc = 0;
+	char *data = NULL;
+
+	if (tid == NULL) {
+		set_err_msg(resp, TGT_ERR_INVALID_PARAM,
+			"tid param not given");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	data = ha_get_data(reqp);
+	if (data == NULL) {
+		set_err_msg(resp, TGT_ERR_NO_DATA,
+			"json config not given");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	json_error_t error;
+	json_auto_t *root = json_loads(data, 0, &error);
+
+	free(data);
+	if (root == NULL) {
+		set_err_msg(resp, TGT_ERR_INVALID_JSON,
+			"json config is incorrect");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	json_t *tname = json_object_get(root, "TargetName");
+	if (!json_is_string(tname)) {
+		set_err_msg(resp, TGT_ERR_INVALID_TARGET_NAME,
+			"TargetName is not string");
+		return HA_CALLBACK_CONTINUE;
+	}
+
+	memset(cmd, 0, sizeof(cmd));
+	snprintf(cmd, sizeof(cmd), "tgtadm --lld iscsi --mode target --op new"
```

Comment:
You might want to check return value of snprintf. If thing u wanted in string wrapped.

Reply:
done.